src/relay/op/memory/memory.cc

```diff
@@ -360,12 +360,26 @@ bool ShapeFuncRel(const Array<Type>& types, int num_inputs, const Attrs& attrs,
   auto tuple = TupleType(func_type->arg_types);
   auto in_types = FlattenTupleType(tuple);
   auto out_types = FlattenTupleType(func_type->ret_type);
+  int num_types = 0;
```

Comment:
Can you use the `FlattenTupleType` helper instead of manually processing them like this? this won't work for nesting and is pretty close to the same tuple processing code written everywhere.

Reply:
The problem here is that we need to restore is_input to make it correspond to the flattened input types. However, is_input is created in memory alloc pass, it is in flatten pattern where a tuple input just get one single number instead of tuple of number. As a result we cannot use similar way of ```FlattenTupleType```. This also makes it more complicated for handling nested tuple as input.

Reply:
@jroesch Improved implementation with FlattenTupleType and it can handle nested tuple type now. Also update test case to include nested tuple.